### PR TITLE
Remove the akka agent alterOff and sendOff

### PR DESCRIPTION
**Pull request by QuarpT**
The purpose of this commit is to remove the akka agent 'alterOff' method, used for running blocking operations in futures. It has been replaced by a thread pool execution context. 

Refactoring was required to get a reference to Play's ActorSystem, only accessible via dependency injection. This meant converting lots of Scala singletons to injected classes.

Also removed a circular dependency, DfpApi -> DataMapper -> CustomFieldService -> CustomFieldAgent -> DfpApi

## What does this change?

Remove akka agent 'alterOff' method

## What is the value of this and can you measure success?

Part of the Scala 2.12 upgrade, preparing to remove replace Akka Agent with box.

## Does this affect other platforms - Amp, Apps, etc?

Admin app

## Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

no

## Tested in CODE?

Will do

### CODE REVIEW

File: admin/app/dfp/AdvertiserAgent.scala [hunk 1]
```diff
@@ -2,17 +2,17 @@ package dfp
...
+class AdvertiserAgent(blockingOperations: BlockingOperations, dfpApi: DfpApi) {
 
   private lazy val cache = AkkaAgent(Seq.empty[GuAdvertiser])
 
   def refresh()(implicit executionContext: ExecutionContext): Future[Seq[GuAdvertiser]] = {
-    cache alterOff { oldData =>
-      val freshData = DfpApi.getAllAdvertisers
-      if (freshData.nonEmpty) freshData else oldData
+    blockingOperations.executeBlocking(dfpApi.getAllAdvertisers).map { freshData =>
+      if (freshData.nonEmpty) freshData else cache.get
```
Comment (TBonnin) [inline admin/app/dfp/AdvertiserAgent.scala]: We are not updating the cache anymore here.

File: admin/app/dfp/CreativeTemplateAgent.scala [hunk 2]
```diff
@@ -2,17 +2,17 @@ package dfp
...
+class CreativeTemplateAgent(blockingOperations: BlockingOperations, dfpApi: DfpApi) {
 
   private lazy val cache = AkkaAgent(Seq.empty[GuCreativeTemplate])
 
   def refresh()(implicit executionContext: ExecutionContext): Future[Seq[GuCreativeTemplate]] = {
-    cache alterOff { oldData =>
-      val freshData = DfpApi.readActiveCreativeTemplates()
-      if (freshData.nonEmpty) freshData else oldData
+    blockingOperations.executeBlocking(dfpApi.readActiveCreativeTemplates()).map { freshData =>
+      if (freshData.nonEmpty) freshData else cache.get()
```
Comment (TBonnin) [inline admin/app/dfp/CreativeTemplateAgent.scala]: Same as above

File: admin/app/dfp/DataAgent.scala [hunk 3]
```diff
@@ -10,24 +11,28 @@ trait DataAgent[K, V] extends Logging with implicits.Strings {
...
+        val duration = System.currentTimeMillis - start
+        log.info(s"Loading DFP data (${freshData.keys.size} items}) took $duration ms")
+        DataCache(freshData)
+      case Success(_) =>
+        log.error("No fresh data loaded so keeping old data")
+        cache.get
+      case Failure(e) =>
+        log.error("Loading of fresh data has failed.", e)
+        cache.get
     }
```
Comment (TBonnin) [inline admin/app/dfp/DataAgent.scala]: The cache is not being updated with the new value here neither

File: admin/app/dfp/OrderAgent.scala [hunk 4]
```diff
@@ -2,17 +2,17 @@ package dfp
...
+class OrderAgent(blockingOperations: BlockingOperations, dfpApi: DfpApi) {
 
   private lazy val cache = AkkaAgent(Seq.empty[GuOrder])
 
   def refresh()(implicit executionContext: ExecutionContext): Future[Seq[GuOrder]] = {
-    cache alterOff { oldData =>
-      val freshData = DfpApi.getAllOrders
-      if (freshData.nonEmpty) freshData else oldData
+    blockingOperations.executeBlocking(dfpApi.getAllOrders).map { freshData =>
+      if (freshData.nonEmpty) freshData else cache.get
```
Comment (TBonnin) [inline admin/app/dfp/OrderAgent.scala]: Same here. We are not updating the cache anymore

File: admin/conf/application.conf [hunk 5]
```diff
@@ -22,8 +22,17 @@ akka {
...
     }
   }
+
+  blocking-operations {
+    executor = "thread-pool-executor"
+    throughput = 10
+    thread-pool-executor {
+      fixed-pool-size = 128
+    }
+  }
```
Comment (TBonnin) [inline admin/conf/application.conf]: What's the reasoning behind using a fixed size thread-pool-executor dispatcher?

**Review by TBonnin — CHANGES_REQUESTED** [review]

File: admin/app/dfp/DataAgent.scala (hunk 3 above)
Comment (QuarpT) [reply admin/app/dfp/DataAgent.scala]: Fixed

File: admin/conf/application.conf (hunk 5 above)
Comment (QuarpT) [reply admin/conf/application.conf]: It's ideal for blocking operations

**Review by TBonnin — APPROVED** [review]

**Pull request merged**